bridge/server/user/oauth-routes.ts

```diff
@@ -0,0 +1,149 @@
+import { Request, Response, Router } from 'express';
+import { authenticateSession, getLogoutHint, isAuthenticated, removeSession } from './session';
+import { BaseClient, errors, generators, TokenSet } from 'openid-client';
+
+const prefixPath = process.env.PREFIX_PATH;
+const codeVerifiers: { [state: string]: { codeVerifier: string; nonce: string; expiresAt: number } } = {};
+
+/**
+ * Build the root path. The exact path depends on the deployment & PREFIX_PATH value
+ *
+ * If PREFIX_PATH is defined, root location is set to <PREFIX_PATH> or <PREFIX_PATH>/bridge for production environments. Otherwise, root is set to / .
+ *
+ * Redirection to / will be either handled by Nginx (ex:- generic keptn deployment) OR the Express layer (ex:- local bridge development).
+ */
+function getRootLocation(): string {
+  if (prefixPath !== undefined) {
+    return process.env.NODE_ENV === 'production' ? `${prefixPath}/bridge` : prefixPath;
+  }
+
+  return '/';
+}
+
+function oauthRouter(client: BaseClient, redirectUri: string, reduceRefreshDateSeconds: number): Router {
+  const router = Router();
+
+  /**
+   * Router level middleware for login
+   */
+  router.get('/login', async (req: Request, res: Response) => {
+    const codeVerifier = generators.codeVerifier();
+    const codeChallenge = generators.codeChallenge(codeVerifier);
+    const nonce = generators.nonce();
+    const state = generators.state();
+    codeVerifiers[state] = { codeVerifier, nonce, expiresAt: new Date().getTime() + 5 * 60_000 }; // expires in 5 minutes
+
+    const authorizationUrl = client.authorizationUrl({
+      scope: 'openid',
+      state,
+      nonce,
+      code_challenge: codeChallenge,
+      code_challenge_method: 'S256',
+    });
+
+    res.redirect(authorizationUrl);
+    return res;
+  });
+
+  /**
+   * Router level middleware for redirect handling
+   */
+  router.get('/oauth/redirect', async (req: Request, res: Response) => {
+    const params = client.callbackParams(req);
+
+    if (!params.code || !params.state) {
+      return res.redirect(getRootLocation());
+    }
+
+    const verifiers = codeVerifiers[params.state];
+    if (verifiers) {
+      delete codeVerifiers[params.state];
+      params.state = undefined;
+    } else {
+      return res.render('error', {
```

Comment:
In case ao an invalid state, I think it should be 401 response. e.g.: https://developers.google.com/identity/protocols/oauth2/openid-connect#confirmxsrftoken

Reply:
We decided to stay with the error page. Text is adjusted